List item marker should be rendered as the first child of the list item when `list-style-position: inside` is set

AlexsanderDamaceno opened the pull request:
#### 7a8d12f962c7ea8d3ef3702c867f540e5c5d38e1
<pre>
List item marker should be rendered as the first child of the list item when `list-style-position: inside` is set
<a href="https://bugs.webkit.org/show_bug.cgi?id=227000">https://bugs.webkit.org/show_bug.cgi?id=227000</a>

Reviewed by Alan Baradlay.

As stated by <a href="https://developer.mozilla.org/en-US/docs/Web/CSS/list-style-position">https://developer.mozilla.org/en-US/docs/Web/CSS/list-style-position</a>

&quot;If a block element is the first child of a list element declared as `list-style-position: inside`,
then the block element is placed on the line after the marker-box.&quot;

In the current implementation, we render the marker on the same first line
as the content flow when the first child is a block element. This patch
changes the behavior to render the marker in the first line of the
list item, instead of the first line of the content when `list-style-position:inside` is present.

this align behavior with other browsers.
* LayoutTests/TestExpectations
* LayoutTests/fast/lists/list-marker-style-inside.html: Added.
* LayoutTests/fast/lists/list-marker-style-inside-expected.html: Added.
* LayoutTests/fast/repaint/list-marker-expected.txt: Added.
* LayoutTests/platform/mac/fast/repaint/list-marker-expected.txt: Removed.
* LayoutTests/platform/win/fast/repaint/list-marker-expected.txt: Removed.

* Source/WebCore/rendering/updating/RenderTreeBuilderList.cpp:
(WebCore::parentCandidateForMarker):

Canonical link: <a href="https://commits.webkit.org/296451@main">https://commits.webkit.org/296451@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43fb2384f5c24a713c65771c35bf5bab40e95c13

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107841 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27509 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17927 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113055 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58365 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109807 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28204 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36052 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81851 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110786 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22339 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97158 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62272 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21760 "Passed tests") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK1-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57808 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91691 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15356 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116176 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34910 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25691 "Found 1 new test failure: fast/webgpu/nocrash/fuzz-291180.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90883 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35286 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93434 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90672 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35573 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30660 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17532 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34814 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40366 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34556 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37917 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36218 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/81aae342f3a95a89214cd1e62da750859e7a5eaa. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b8f1ab45df6ea8efea0da034fd8543732e2911e3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5065d8d4a4ed15d3965c44b2dc48c6d632c4741e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ff694b56b8c800ff5ad519d82bab3853571e50ca. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6cc9326a48deae6e1bea4a22ba4607d18140c1ca. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6609e37db7e771ccc83cdb85d14a732505937412. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/dfc272c1b45a9d87233a48f2169dc5829d5f9d07. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8b2edb12b8ea8220d391907438543ff5453f9477. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/927ceba20f2b28b3f860c01070064fcbe539f886. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9566add6937865c979e876244661fdceb4b629db. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c4693658fcd7ad07260a08978b14d9987b39d890. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6fba7789698a6ba3fbd0f3f1c47caaf7b7aa5c4a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/717e6f59be92c682507768a7b67ba8230e501c99. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d658659d030bed8f0f0f4be6c3a32da1120b3581. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

alanbaradlay commented on Source/WebCore/rendering/updating/RenderTreeBuilderList.cpp:
```diff
@@ -49,6 +49,7 @@ static bool generatesLineBoxesForInlineChild(RenderBlock& current, RenderObject*
 static RenderBlock* getParentOfFirstLineBox(RenderBlock& current, const RenderObject& marker)
```
I recently landed a few PRs around this code so I think now you can move this logic to `parentCandidateForMarker` something along the lines of

 ```
   if (marker.isInside()) {
        if (firstChild->isBlockLevelBox())
          return &listItemRenderer;
       return findParentOfEmptyOrFirstLineBox(listItemRenderer, marker).first;
    }

```

(with some function signature changes (`RenderBlock& blockContainer` -> `RenderListItem& listItemRenderer` and grabbing the first child as RenderElement)

or you can simply land your change and I'll move it over to `parentCandidateForMarker`. I am fine with either.

alanbaradlay commented on LayoutTests/fast/repaint/list-marker-expected.txt:
```diff
@@ -0,0 +1,48 @@
+layer at (0,0) size 800x600
```
I don't think this is needed since you have a ref test setup here.

alanbaradlay replied:
oh, my bad.

alanbaradlay replied:
but this is a new file, I am confused.

alanbaradlay replied:
ok, makes sense. if all platforms match, you can remove the platform specific tree dumps.

alanbaradlay commented on LayoutTests/fast/lists/list-marker-style-inside-ref.html:
```diff
@@ -0,0 +1,6 @@
+<!doctype html>
```
-refs are specific to WPT tests. They don't work in LayoutTests/fast.

alanbaradlay commented on LayoutTests/fast/lists/list-marker-style-inside-expected.html:
```diff
@@ -0,0 +1,6 @@
+  <li style="list-style-position:inside">
+    <div style="border:1px solid">The bullet should be above this line.</div>
+  </li>
+</ul>
```
hmm, this looks exactly like the test case.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e1ff079c2ea6ad9f3c34a06c545d6969f359abd2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3c3d18d6336515aaec29d15eff1668f849f6f6df. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3c3d18d6336515aaec29d15eff1668f849f6f6df. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7ef4fd48a15744c05169e732d7268059ae25db87. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7ef4fd48a15744c05169e732d7268059ae25db87. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/db140aae154c05a93ec9e11495d4f7293cc307b7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/db140aae154c05a93ec9e11495d4f7293cc307b7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/48bf0de43a6e4337d35b74d8a0453398bece46a8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/016c74d9f597307e59f1b588dbffdd460895cb0c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1381ebdfe41d470f63fca708ea8c3a41e9dacfeb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

nt1m commented on Source/WebCore/rendering/updating/RenderTreeBuilderList.cpp:
```diff
@@ -80,9 +80,15 @@ static std::pair<RenderBlock*, RenderBlock*> findParentOfEmptyOrFirstLineBox(Ren
+            return &listItemRenderer;
+        return findParentOfEmptyOrFirstLineBox(listItemRenderer, marker).first;
+    }
+    auto [parentCandidate, fallbackParent ] = findParentOfEmptyOrFirstLineBox(listItemRenderer, marker);
```
```suggestion
    auto [parentCandidate, fallbackParent] = findParentOfEmptyOrFirstLineBox(listItemRenderer, marker);
```

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/75a0525bea7d19cc5c9e1fa1dbc694d68aba1ca7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

nt1m reviewed (changes requested):
The reftest failure suggests the bug isn't fixed. See how the testcase from Mats renders in Firefox or Chrome vs. Safari: https://bugs.webkit.org/attachment.cgi?id=431387

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e3e157697c00a1988f5664f2d0db4a14981b3e12. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

alanbaradlay reviewed (approved):
The code change looks ok, please make sure the test case is also ok before landing.

nt1m commented on Source/WebCore/rendering/updating/RenderTreeBuilderList.cpp:
```diff
@@ -80,9 +80,16 @@ static std::pair<RenderBlock*, RenderBlock*> findParentOfEmptyOrFirstLineBox(Ren
 {
-    auto [parentCandidate, fallbackParent ] = findParentOfEmptyOrFirstLineBox(blockContainer, marker);
+    auto* listItemParent = listItemRenderer.parent()->element();
+    if (marker.isInside() && listItemParent && isHTMLListElement(*listItemParent)) {
```
I'm surprised this is specific to HTML `<li>` elements, and not simply to `display: list-item` (which is implied here).

@Loirooriol @fantasai Can you double check if this is correct?

nt1m replied:
I think you just be able to drop ` && listItemParent && isHTMLListElement(*listItemParent)` in that case. I think it's implied that the parent is a list item in this context

Loirooriol replied:
I don't understand @AlexsanderDamaceno's comment above, but I agree the logic shouldn't depend on whether the list item is a `<li>` or not.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/79f53761e0ee5380a1b6e2527d99ec48ce5d2f8e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/79f53761e0ee5380a1b6e2527d99ec48ce5d2f8e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f957a7fb16f588e30a623456106f36ee2ed87830. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f957a7fb16f588e30a623456106f36ee2ed87830. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/43fb2384f5c24a713c65771c35bf5bab40e95c13. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/43fb2384f5c24a713c65771c35bf5bab40e95c13. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46031

Ahmad-S792 commented on LayoutTests/TestExpectations:
```diff
@@ -5273,6 +5273,7 @@ imported/w3c/web-platform-tests/css/css-images/tiled-conic-gradients.html [ Skip
 
 # Untriaged counter failures
 webkit.org/b/214457 imported/w3c/web-platform-tests/css/css-lists/change-list-descendant-display.html [ ImageOnlyFailure ]
+imported/w3c/web-platform-tests/css/css-lists/change-list-style-position-002.html [ ImageOnlyFailure ]
```
Is it significant difference because else it would be regression in this test. Don't know whether it is OK or not. @alanbaradlay - might have some insight.

Ahmad-S792 commented on LayoutTests/fast/lists/list-marker-style-inside.html:
If I am not mistaken - WPT test case for these cases -> https://wpt.fyi/results/css/CSS2/lists?label=master&label=experimental&aligned&q=safari%3Afail (list-style-position-024.xht & list-style-position-023.xht).

AlexsanderDamaceno committed:
List item marker should be rendered as the first child of the list item when `list-style-position: inside` is set
https://bugs.webkit.org/show_bug.cgi?id=227000

Reviewed by Alan Baradlay.

As stated by https://developer.mozilla.org/en-US/docs/Web/CSS/list-style-position

"If a block element is the first child of a list element declared as `list-style-position: inside`,
then the block element is placed on the line after the marker-box."

In the current implementation, we render the marker on the same first line
as the content flow when the first child is a block element. This patch
changes the behavior to render the marker in the first line of the
list item, instead of the first line of the content when `list-style-position:inside` is present.

this align behavior with other browsers.
* LayoutTests/TestExpectations
* LayoutTests/fast/lists/list-marker-style-inside.html: Added.
* LayoutTests/fast/lists/list-marker-style-inside-expected.html: Added.
* LayoutTests/fast/repaint/list-marker-expected.txt: Added.
* LayoutTests/platform/mac/fast/repaint/list-marker-expected.txt: Removed.
* LayoutTests/platform/win/fast/repaint/list-marker-expected.txt: Removed.

* Source/WebCore/rendering/updating/RenderTreeBuilderList.cpp:
(WebCore::parentCandidateForMarker):

Canonical link: https://commits.webkit.org/296451@main
